nav2_smac_planner/src/node_hybrid.cpp

```diff
@@ -314,7 +321,7 @@ float NodeHybrid::getTraversalCost(const NodePtr & child)
 
   float travel_cost = 0.0;
   float travel_cost_raw =
-    NodeHybrid::travel_distance_cost +
+    NodeHybrid::travel_distance_cost * motion_table.travel_distance_reward +
```

Comment:
Can you give a bit more of an expanded explanation of this and what values you find are "reasonable"? Was the drop in speed so significant we need new flags to artificially speed it up? Where's that compute time going to, specifically?  

What's the expansion pattern look like now?

Reply:
> Was the drop in speed so significant we need new flags to artificially speed it up?

It depends on situation and other-than-costmap penalties but in some situations I recorded computation time rise from 114ms to 583ms so it can be quite significant (although ["turning traversal penalties off"](https://github.com/ros-planning/navigation2/issues/2498#issuecomment-966579662) can make the difference smaller).

> Where's that compute time going to, specifically?

There happen to be more node expansions in the main A* search. Reason is that since 2D heuristic search now returns an optimal 2D path cost, the resulting H value is lower than the H value of a sub-optimal path found by the inadmissible search. Not by much but even a small difference introduces a large difference in expansions count (as you must have already experienced too [when retuning penalties](https://github.com/ros-planning/navigation2/issues/2498#issuecomment-966579662)). Here you can see a situation where decreasal of H value by 1.33% (largest H value: 1293 -> 1276) raises the number of expansions (the rainbow-colored band) by 561% (46984 -> 263829).
![Screenshot from 2022-01-12 17-26-39](https://user-images.githubusercontent.com/6342166/149757674-aa8e7120-408b-44e7-b6b7-f2990934f9de.png)
![Screenshot from 2022-01-12 17-28-16](https://user-images.githubusercontent.com/6342166/149757683-dfd03b82-13b2-4ba6-b946-cf42bbf1e848.png)

> Can you give a bit more of an expanded explanation of this?

The reason of the above (H-value/expansions-count) gain being so large is that the search wants to find an optimal path and thus every little other-than-costmap traversal penalty and every little deviation from the 3D path from the 2D optimal path makes the search algorithm ask "Could a change made earlier along the path make the current deviation/penalty smaller?". Therefore in that situation it usually expands a large number of nodes all along the path trying to find the answer (~a better path). However, in a vast majority of situations the solution is somewhere near the "problematic node" and therefore a question arises: "Is it worth to look for an answer so far backwards?". Introducing retrospective penalty gives an option to the search to ask such question and thus spare time not expanding a large number of nodes which have just a very small chance to improve the path.

Of course, it doesn't affect the ability of finding a path, it just gradually prioritizes "later/longer" search branches before "earlier/shorter" ones.

Technically it doesn't increase the cost of all earlier nodes, instead it decreases the cost of newly added ones.

> And what values you find are "reasonable"?

It depends on how much speedup you need and how difficult your map is.

Of course, when this penalty is turned on, the search loses any guarantees for the path to be optimal. If you use a Reeds-Shepp model and your goal is e.g. to reverse into a long narrow passage, inside which it's difficult to rotate (although possible but very problematic and turn should be chosen just in case there are no other options), you will want a value of 0.0, because whenever planner finds the robots should have turned backwards, it needs to go back all along the path to plan a rotation before the entrance to the passage. If a retrospective penalty was used, a later turn inside the passage could be prefered and robot could get stuck doing the dangerous rotation.

If situations like the above one are not present in your environment/use-case, you can penalize the retrospective a bit. According to our tests, value 0.01 decreased number of expansions of the example above down to 35044 expansions (less than with the original inadmissible search), while still dealing pretty well with the passage reversal situation described in the paragraph above.
![Screenshot from 2022-01-13 08-18-36](https://user-images.githubusercontent.com/6342166/149764451-a8be2322-5bc0-4e7e-9cba-0e9d6f6d95fa.png)

Following is an illustrative example with value 0.2. You can see how the rainbow-colored band expands only near the problematic parts. The search still deals very well with problems along the path, however the value is maybe quite marginal, making it possible in some situations to prefer paths which are problematic to execute.
![Screenshot from 2022-01-13 09-00-25](https://user-images.githubusercontent.com/6342166/149765532-efcd9b06-dd56-4b71-afad-783730e3839c.png)

Reply:
> in some situations I recorded computation time rise from 114ms to 583ms so it can be quite significant  ... raises the number of expansions (the rainbow-colored band) by 561% (46984 -> 263829).

:grimacing: yeeek! When adding in the non-0 values, does this get back in line with the ~114 ms number? I see the expansions go back in line, just checking that expansions + new H computation still is in the same ballpark. 

A value of 0.2 would be 20% off which would seems high. It doesn't surprise me that you see some issues if you discount expansions by 20% that it would cause some non-optimal issues. It does surprise me that even just 1% bounds things so quickly. 

If you were to pick a "reasonable" default value, what would it be? Sounds like between 0.01 - 0.20.

Reply:
> When adding in the non-0 values, does this get back in line with the ~114 ms number?

Did you mean non-0 value of `retrospective_penalty`? Yes, it brings the computation down to around 120 ms or even lower

> It does surprise me that even just 1% bounds things so quickly

As I said, there was around 1.33% decrease of heuristic values with the new admissible computation so 1% brings the expansion number back (+ even lower since the retrospective penalization here is continuous, taking effect all along path, which makes it more effective than the one introduced by an inadmissible heuristic, which was error-driven and thus occurred presumably mostly around obstacle corners etc.).

> If you were to pick a "reasonable" default value, what would it be?

For me it is OK to be 0.01 (you know I like "more optimal paths and longer planning time" more than "less optimal paths and shorter planning time") or even lower. But since this is just one test/comparison, we could use some reserve, e.g. 0.015. But I would like you to find the proper value because our priorities differ here.

Reply:
Got it - OK. If you could set the value in your PR to have a default value in the plugins of `0.015` that would be great. Since that puts thing back in line, **I'd support removing my existing implementation** here. Which means you can delete that stuff and change this to being the only obstacle heuristic.

Reply:
I retested the planner with more paths and decided to use 0.025 as default to keep expansions number < 50000 (=> planning time < 120ms) for all 50 m long paths I tested with...